core/trino-main/src/test/java/io/trino/sql/TestExpressionInterpreter.java

```diff
@@ -1397,6 +1421,61 @@ public void testLikeOptimization()
     }
 
+    @Test
+    public void testLikeCharOptimization()
+    {
+        // constant literal pattern of length shorter than value length
+        assertOptimizedEquals("unbound_char LIKE 'abc'", "false");
+        assertOptimizedEquals("unbound_char LIKE 'abc' ESCAPE '#'", "false");
+        assertOptimizedEquals("unbound_char LIKE 'ab#_' ESCAPE '#'", "false");
+        assertOptimizedEquals("unbound_char LIKE 'ab#%' ESCAPE '#'", "false");
+        assertOptimizedEquals("CAST(unbound_char AS char(4)) LIKE 'abc'", "false");
+        assertOptimizedEquals("CAST(unbound_char AS char(4)) LIKE 'abc' ESCAPE '#'", "false");
+
+        // constant non-literal pattern of length shorter than value length
+        assertOptimizedEquals("unbound_char LIKE 'ab_'", "unbound_char LIKE 'ab_'");
+        assertOptimizedEquals("unbound_char LIKE 'ab%'", "unbound_char LIKE 'ab%'");
+        assertOptimizedEquals("unbound_char LIKE 'ab%' ESCAPE '#'", "unbound_char LIKE 'ab%' ESCAPE '#'");
+
+        // constant literal pattern of length equal to value length
+        assertOptimizedEquals("CAST(unbound_char AS char(4)) LIKE 'abcd'", "CAST(unbound_char AS char(4)) = CAST('abcd' AS char(4))");
+        assertOptimizedEquals("CAST(unbound_char AS char(4)) LIKE 'abcd' ESCAPE '#'", "CAST(unbound_char AS char(4)) = CAST('abcd' AS char(4))");
+        assertOptimizedEquals("CAST(unbound_char AS char(4)) LIKE 'jaźń'", "CAST(unbound_char AS char(4)) = CAST('jaźń' AS char(4))");
+        assertOptimizedEquals("CAST(unbound_char AS char(4)) LIKE 'ab#_' ESCAPE '#'", "false");
+        assertOptimizedEquals("CAST(unbound_char AS char(4)) LIKE 'ab#%' ESCAPE '#'", "false");
+
+        // constant non-literal pattern of length equal to value length
+        assertOptimizedEquals("CAST(unbound_char AS char(4)) LIKE 'ab#%' ESCAPE '\\'", "CAST(unbound_char AS char(4)) LIKE 'ab#%' ESCAPE '\\'");
+
+        // constant pattern of length longer than value length
+        assertOptimizedEquals("CAST(unbound_char AS char(4)) LIKE 'abcde'", "false");
+        assertOptimizedEquals("CAST(unbound_char AS char(4)) LIKE 'abcde' ESCAPE '#'", "false");
+        assertOptimizedEquals("CAST(unbound_char AS char(4)) LIKE '#%a#%b#%c#%d#%' ESCAPE '#'", "false");
+
+        // constant non-literal pattern of length longer than value length
+        assertOptimizedEquals("CAST(unbound_char AS char(4)) LIKE '%a%b%c%d%'", "CAST(unbound_char AS char(4)) LIKE '%a%b%c%d%'");
+        assertOptimizedEquals("CAST(unbound_char AS char(4)) LIKE '%a%b%c%d%' ESCAPE '#'", "CAST(unbound_char AS char(4)) LIKE '%a%b%c%d%' ESCAPE '#'");
+
+        // without explicit CAST on value, constant pattern of equal length
+        assertOptimizedEquals(
+                "unbound_char LIKE CAST(CAST('abc' AS char( " + TEST_CHAR_TYPE_LENGTH + ")) AS varchar(" + TEST_CHAR_TYPE_LENGTH + "))",
+                "unbound_char = CAST('abc              ' AS char(17))");
```

Comment:
why trailing spaces?

Reply:
this is actually redundant, fixed.

the bigger problem is that `assertOptimizedEquals` runs optimize for both input and expected, so this assertion is very imprecise.
i don't know how to fix that